### PR TITLE
Adding doubtlab from PyPI

### PR DESCRIPTION
**Pull request by sugatoray**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/main/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged). `NA`
- [ ] If static libraries are linked in, the license of the static library is packaged. `NA`
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team. `NA`

### CODE REVIEW

**Comment by sugatoray** [comment]
@conda-forge/help-python I believe the PR is ready for review now. Thank you.

**Comment by sugatoray** [comment]
Related Issue in github/doubtlab repo: https://github.com/koaning/doubtlab/issues/14

File: recipes/doubtlab/meta.yaml
Comment (BastianZim) [inline recipes/doubtlab/meta.yaml:58]: Is there a specific reason you're not using pip check? It's generally good practice to make sure that the reqs are up-to-date.

File: recipes/doubtlab/meta.yaml
Comment (BastianZim) [inline recipes/doubtlab/meta.yaml:27]: Conda-forge has a bot that will automatically create a PR for a new release every time one is pushed to PyPI, you won't need to update the `sha256` and version yourself. The only thing you'll need to pay attention to is the requirements and license (If you change that).

Comment (BastianZim) [reply recipes/doubtlab/meta.yaml:27]: Example: https://github.com/conda-forge/lark-feedstock/pull/2

Comment (sugatoray) [reply recipes/doubtlab/meta.yaml:27]: @BastianZim Thank you, for clarifying that and reviewing. Should I remove the `fetch_sha256.sh` then, or is it okay to leave it there? What do you suggest?

Also, (assuming you meant the `@auto-tick` bot) is there some script that could extract the dependency-versions directly from the package and update the requirements?

Something that I am not clear about is, given I have some shell script to extract the requirements from the PyPI package itself, how can I update that in the `meta.yaml` file?

File: recipes/doubtlab/meta.yaml
Comment (sugatoray) [reply recipes/doubtlab/meta.yaml:58]: Ok. I saw it was optional and so to keep things simple initially, I chose to comment out those lines. Will uncomment those lines and update the PR. Thank you.

Comment (BastianZim) [reply recipes/doubtlab/meta.yaml]: Yes, it's optional but it's just a nice extra "set of eyes". It doesn't take much CI time so it's fine to include it.

File: recipes/doubtlab/meta.yaml
Comment (BastianZim) [reply recipes/doubtlab/meta.yaml]: I would remove it and let the bot handle it.

Unfortunately not. You can use Grayskull for this but this is something you'll need to manually double-check. The bot-PR will also include a check but it's experimental and not guaranteed to work 100%.

Basically, a version update works like this:  

1. A new version is published on PyPI.
2. The conda-forge bot picks this up (Note, this can take a couple of hours sometimes. The status page for this is https://conda-forge.org/status/#version_updates
3. The bot opens a new PR, updating the version and `sha256`.
4. The maintainer verifies that no requirements changed and that the license and, if provided, URLs are still the same.
5. The maintainer merges the PR.

I'm not sure what you mean by updating the meta.yaml file. This is done via a normal PR (Please use a personal fork for this, as each push is uploaded to Anaconda.org, no matter which branch).

**Comment by BastianZim** [comment]
The new error is due to a dependency, where the dependencies weren't correctly added, this has nothing to do with your PR.
Depending on how much you want to get involved, this can be solved by either adding it here or updating the feedstock of the dependency, which is https://github.com/conda-forge/cleanlab-feedstock

The latter is preferred from a conda-forge perspective as feedstocks should always be up-to-date but that means that you would maintain an additional feedstock, not sure if you have enough time for this and are willing to.

**Comment by sugatoray** [comment]
@BastianZim It threw an error while running `pip check`. [see here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=412080&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1291)

It seems `tqdm` is needed, but is not installed. So, where should I install that `tqdm`? Add it to `test.requires`?

```sh
export PREFIX=/home/conda/staged-recipes/build_artifacts/doubtlab_1636999752641/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho
export SRC_DIR=/home/conda/staged-recipes/build_artifacts/doubtlab_1636999752641/test_tmp
import: 'doubtlab'
import: 'doubtlab'
+ pip check
cleanlab 1.0 requires tqdm, which is not installed.
Tests failed for doubtlab-0.1.1-pyhd8ed1ab_0.tar.bz2 - moving package to /home/conda/staged-recipes/build_artifacts/broken
WARNING:conda_build.build:Tests failed for doubtlab-0.1.1-pyhd8ed1ab_0.tar.bz2 - moving package to /home/conda/staged-recipes/build_artifacts/broken
WARNING conda_build.build:tests_failed(2955): Tests failed for doubtlab-0.1.1-pyhd8ed1ab_0.tar.bz2 - moving package to /home/conda/staged-recipes/build_artifacts/broken
TESTS FAILED: doubtlab-0.1.1-pyhd8ed1ab_0.tar.bz2
##[error]Bash exited with code '1'.
Finishing: Run docker build
```

**Comment by BastianZim** [comment]
See my comment [here](https://github.com/conda-forge/staged-recipes/pull/16940#issuecomment-969192905) but this should be added in `run`.

**Comment by sugatoray** [comment]
> See my comment [here](https://github.com/conda-forge/staged-recipes/pull/16940#issuecomment-969192905) but this should be added in `run`.

Thank you. I added `tqdm` to `test.requires` and it passed all the CI's. About the amount of involvement/time availability, I would (for now) just add `tqdm` here. And perhaps later on consider about further involvement.

**Comment by BastianZim** [comment]
Yes, but it will be needed in `run` as otherwise, users won't be able to use the package. Test requirements are generally only packages like `pytest` etc. that aren't needed when using the package, which is not the case here.

I have created https://github.com/conda-forge/cleanlab-feedstock/pull/3 Once that is merged, you should be able to remove `tqdm` again.

**Comment by sugatoray** [comment]
@BastianZim Thank you, for opening the PR. I have added `tqdm` in `requirements.run`.

**Comment by BastianZim** [comment]
Looks good, thanks!

Let me ping the review team then for a final look.

@conda-forge/help-python

File: recipes/doubtlab/meta.yaml
Comment (sugatoray) [reply recipes/doubtlab/meta.yaml]: Thank you. I removed the `fetch_sha256.sh` file.

Comment (sugatoray) [reply recipes/doubtlab/meta.yaml]: Thank you, for explaining the version-update process-flow.

**Comment by sugatoray** [comment]
@BastianZim Thank you, for all the help and your reviews. I have made a few more changes to comments (mostly) to keep note of the things I learned in the process. 

A quick question: what is the typical wait-time for the `@conda-forge/help-python` to respond/merge this PR?

**Comment by BastianZim** [comment]
Depends really on how busy they are or if someone’s already around. Sometimes it’s a couple of minutes, sometimes it's a couple of days. But they’re generally very quick.

**Comment by BastianZim** [comment]
As everyone here is a volunteer it's unfortunately impossible to give a better timeframe.

**Comment by sugatoray** [comment]
> As everyone here is a volunteer it's unfortunately impossible to give a better timeframe.

Thank you. I get it. After all it's open-source. I am also volunteering on this conda-forge recipe PR. Fortunately, I was able to put some more time on this today than I would be normally. 

I will wait for them to respond.

**Comment by sugatoray** [comment]
@BastianZim This was a rather simple exercise, given the library has very few dependencies. But you could also have libraries that may offer many `extras` to install with pip (for supporting added functionalities).

```sh
pip install "somelib[extra1,extra2,extra3]"
```

Is there any best-practice to handle such scenarios? Or, am I supposed to just add all the extras together into the same feedstock?

**Comment by BastianZim** [comment]
Unfortunately, that's a situation that does not (yet) have a clear answer as that functionality is not provided by conda. There are a couple of approaches used by the community though:

1. [`run_constrained`](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#run-constrained)
  `run_constrained` basically acts like an `if installed`. What it does, when being added to a recipe, is restrict a package without actually installing it. For example, if you `run_constrained` _requests >=3_, the package will not be installed as a dependency but if someone installs _requests_ in the same environment where your package is installed, conda will make sure that only versions _>=3_ are installed.
2. Add all dependencies
  This is the easiest approach because you just add everything but isn't always the best one. One example, where this is bad, would be if the dependency has a much more limited license than your package.
_Example_: `or-tools` has a very unrestricted license but can be shipped with SCIP that forces you to buy a commercial license in some cases. Another dependency that can be used instead also has a very unrestricted license so it wouldn't make sense to force both on the user.
Another example is the size of the dependencies: Some packages are very lean themselves but can be run with other huge packages that maybe not everybody wants to install if they don't need their functionality.
3. Create [multiple outputs](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#outputs-section)
  Conda allows you to specify multiple outputs in a recipe. This means that you only have one source and one recipe but multiple packages are created.
This can be used to create a setup, where you build the actual package with minimal dependencies and then create empty packages that depend on the initial package + the optional dependencies. These extra outputs can then be named _package-name_-_option_.

**Comment by sugatoray** [comment]
@BastianZim Thank you, for explaining with such care. I like approach 3. Will try to use them for packaging another library I started working on.

**Comment by BastianZim** [comment]
Happy to. 

A good example for 3. would be https://github.com/conda-forge/coin-or-cbc-feedstock/blob/master/recipe/meta.yaml by the way. 
Here the goal is to have a package with two different names but the same content (`coincbc` was the old name and it's now included to not break any environments that rely on it), but this can also be used for a package with the same name but different dependencies.

**Comment by sugatoray** [comment]
@BastianZim I used `run_constrained` successfully. But, now am trying to use `output` and I keep getting a "`PYTHON` not found" error. Would it be possible for you to take a look at it?

https://github.com/conda-forge/staged-recipes/pull/16963#issuecomment-972140685

**Comment by BastianZim** [comment]
Gentle ping here @conda-forge/help-python

**Comment by sugatoray** [comment]
Gentle ping here @conda-forge/help-python

**Comment by BastianZim** [comment]
Hmm seems like the bot isn't working. 

@conda-forge/staged-recipes gentle ping here. Also, the bot isn't pinging anymore - is there a new command that should be used?

File: recipes/doubtlab/meta.yaml
Comment (ocefpaf) [inline recipes/doubtlab/meta.yaml:25]: These comments are not really necessary, right? Can we remove them?

File: recipes/doubtlab/meta.yaml
Comment (ocefpaf) [inline recipes/doubtlab/meta.yaml:62]: We probably only need line 62 from these comments.

File: recipes/doubtlab/meta.yaml
Comment (sugatoray) [reply recipes/doubtlab/meta.yaml:9]: @ocefpaf I have removed/updated the comments to another file (`maintainers_notes.md`). This will serve as a single point of entry for sharing notes among the maintainers of this feedstock.

File: recipes/doubtlab/meta.yaml
Comment (sugatoray) [reply recipes/doubtlab/meta.yaml:34]: @ocefpaf Thank you. Suggestion adopted. Same as [**this comment**][#comment].

[#comment]: https://github.com/conda-forge/staged-recipes/pull/16940/files/d8d847786c94ca85203fd2678d1dd187425021a9#r757835861

**Comment by sugatoray** [comment]
@ocefpaf @BastianZim A question for you: recently I found a library (in `conda-forge`), that did not have all the necessary dependencies (as compared to PyPI). So, I made the updates to its feedstock and pushed a PR. The PR ran successfully through all the CI's. But the maintainers have not responded yet (over 7 days). What are the options here? I learned from the maintainers of library `sentence-transformers`, that the feedstock maintainers have previously been slow to respond and update the package version, despite pushing successfully run PRs: https://github.com/conda-forge/sentence-transformers-feedstock/issues/17. I understand that it is open-source and people are trying to do their best, and so, would be great if someone from conda-forge could merge this PR instead. This will add me to the list of maintainers as well.

**Comment by BastianZim** [comment]
Generally, if a maintainer is unresponsive, you can follow the guidelines here: https://conda-forge.org/docs/maintainer/maintainer_faq.html#mfaq-abandoned-feedstock

Since you have already added yourself, you can contact core to have it merged for you, as outlined here: https://conda-forge.org/docs/maintainer/maintainer_faq.html#mfaq-contact-core

File: recipes/doubtlab/meta.yaml
Comment (BastianZim) [reply recipes/doubtlab/meta.yaml:9]: Why do you need these notes? From what I can see, everything except for the TODOs is covered by the docs. Generally, we prefer to only use those and not add instructions in the recipe, if they're not recipe specific as it clutters the repo and might become outdated. If someone is not familiar with the process, the docs will always have the most up-to-date info.

The TODOs can be included as a single line under the respective items.

Comment (ocefpaf) [reply recipes/doubtlab/meta.yaml:9]: Yep. Anything there that is not in our docs should be added to the docs, not to the recipe.

File: recipes/doubtlab/meta.yaml
Comment (BastianZim) [inline recipes/doubtlab/meta.yaml:35]:
```suggestion
```
Has been merged https://github.com/conda-forge/cleanlab-feedstock/pull/3

**Review by BastianZim — CHANGES_REQUESTED** [review]

File: recipes/doubtlab/meta.yaml
Comment (sugatoray) [reply recipes/doubtlab/meta.yaml]: Thank you.

File: recipes/doubtlab/meta.yaml
Comment (sugatoray) [reply recipes/doubtlab/meta.yaml]: Okay. I removed all the comments and the `maintainers_notes.md` file as well. I have kept all my notes in a GitHub gist. And perhaps maintaining them there is better, as in that case, a mere change in notes does not fire a new PR to the feedstock.

File: recipes/doubtlab/meta.yaml
Comment (ocefpaf) [inline recipes/doubtlab/meta.yaml:53]: @koaning are you OK  being a maintainer here?

**Comment by ocefpaf** [comment]
Thanks for the PR @sugatoray and for the awesome review @BastianZim!
We are only waiting for the second maintainer to confirm they are OK to become maintainer here and we can merge this.

File: recipes/doubtlab/meta.yaml
Comment (BastianZim) [reply recipes/doubtlab/meta.yaml:53]: https://github.com/koaning/doubtlab/issues/14#issuecomment-969213611

**Review by BastianZim — APPROVED** [review]

File: recipes/doubtlab/meta.yaml
Comment (sugatoray) [reply recipes/doubtlab/meta.yaml:53]: @ocefpaf I asked @koaning about his preference, before adding him as another maintainer. He had approved of it (please see [here](https://github.com/koaning/doubtlab/issues/14)). Also, since he is the maintainer of the library `doubtlab`, this would mean both PyPI and conda-forge versions will be easier to keep updated.

**Pull request merged**
